contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,656 @@
+	       
+	       Milestone 7: Initial Plugin changes (GCE, AWS & GlusterFS)
+	       
+	       Milestone 8: Disabling of provisioning where volumeMode == Block is not supported
+
+Phase 2:  v1.9
+Feature: Discovery of block devices 
+
+                Milestone 1: Dynamically provisioned PVs to dynamically allocated devices
+
+                Milestone 2: Privileged container concerns    
+		
+		Milestone 3: Flex volume update
+		
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+## Mounter interface proposed design
+# Plugin changes
+## New BlockVolumeMapper interface proposed design
+
+```
+ type BlockVolumeMapper interface {
+       Volume
+       CanBlockMap() error
+       SetUpDevice(podUID types.UID) error
+       SetUpDeviceAt(dir string, podUID types.UID) error
+       GetDevicePath() (string, error)
+       GetVolumeDeviceMapPath(spec *Spec) (string, error)
+ }
+ type BlockVolumeUnmapper interface {
+       Volume
+       TearDownDevice() error
+       TearDownDeviceAt(dir string) error
+       GetVolumeDeviceUnmapPath(spec *Spec) (string, error)
+ }
+```
+## Changes for volume mount points
+
+Currently, a volume which has filesystem is mounted to the following two paths on a kubelet node when the volumes is in-use.
+The purpose of those mount points are that Kubernetes manages volume attach/detach status using these mount points and number
+of references to these mount points.
+
+```
+- Global mount path
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/{volumePluginDependentPath}/
+
+- Volume mount path
+/var/lib/kubelet/pods/{podUID}/volumes/{escapeQualifiedPluginName}/{volumeName}/
+```
+
+Even if the volumeMode is "Block", similar scheme is needed. However, the volume which 
+doesn't have filesystem can't be mounted.
+Therefore, instead of volume mount, we use symbolic link to map raw block device.
+Kubelet creates a new symbolic link under the new global map path when volume is attached to a Pod. Number of symbolic links
+are equal to the number of Pods which uses the same volume. Kubelet needs to manage both creation and deletion of symbolic links
+under the global map path. The name of the symbolic link is same as pod uuid.
+ 
+```
+Global map path for "Block" volumeMode volume
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/pod-uuid1
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/pod-uuid2
+```
+ 
+Plugin creates a symbolic link under the new volume map path. This symbolic link is not used to manage volume attach/detach
+status but is needed to keep compatibility of current scheme.
+
+```
+Volume map path for "Block" volumeMode volume
+/var/lib/kubelet/pods/{podUID}/volumeDevices/{escapeQualifiedPluginName}/{volumeName}/symlink
+```
```

Comment:
How will these symlinks work on rebooting the node? Lets say I am running 4 pods with attached block devices and node is rebooted. What happens then? in GCE etc, rebooting the node will automatically detach the volume but in EBS volume will still be attached.  In case of a short reboot, the pods will be likely rebooted but old symlinks has to be cleaned up. I would like to see, how this code path will be handled.

Reply:
I am assuming that local block devices are not treated differently from remote block devices like EBS/GCE-PD/Cinder

Reply:
That's a really good point.  Will checking for broken symlink work?

Local and remote are handled the same.

Reply:
There is also a code in ADC that will automatically re-attach devices back when node comes back and pods have not been migrated. I do not think, in that case we can guarantee same device name when a device is reattached. So basically pod UUID will remain the same but device name on node may change..cc @jingxu97

Reply:
>> That's a really good point. Will checking for broken symlink work?

Agree. The /dev/sdX might be changed after rebooting the server, therefore ref count method only checks file existence using os.Lstat or like that, shouldn't check actual device path such as /dev/sdX.

Reply:
Can we require that the volume plugin attaches the device to a unique path?

Reply:
The device path under the /dev is decided by Operation system during boot up, I think plugin can't do anything for this.
If we check symbolic link under the /dev/disk/by-id/, the name of the link is unique and also this symlink is generated during boot up therefore the link always refer correct block device.

If we keep by-id like "{pod uuidX}:by-id', we can refer block device even if the link itself is broken because of changing the device path. I'm currently not sure we need to keep this.

Reply:
So basically symlink that will be created for pod devices will used ids generated inside `/dev/disk/by-id` ?

Reply:
@mtanino @msau42 per our discussion we plan to use the stable name to create the symlink to the device path. In addition, we are exploring how we ensure the symlink only points to the device of origination and can't be subsituted by an additonal device at the same device path

Reply:
@gnufied @erinboyd @jsafrane @msau42 

IIUC, there are two separated problems we are approaching.

1. Stale symlinks across reboot 
2. Silent volume replacement

### Item 1
Here is comment(*1) from @msau42 about node reboot(kubelet restart) scenario. Currently we don't use /dev/disk/by-id for the symlink, but as for the case (a), stale links are overridden by new links during restart of kubelet.
As for case (b), this is still under discussion, we can retrieve volumeSpec from pod device path(*2) using plugin's ConstructVolumeSpec(), therefore, I suppose we may need to recreate symlink from recreated volumeSpec.

```
On a kubelet restart, there's a few scenarios we have to consider when detecting
existing links and whether or not we can trust what it's pointing to:

(a). Pod is still running on the node: The pod will be added to dsw, and in the
reconciler, it will go through setup, then update asw. This should be safe because
we have the full volume spec and can override stale links.

(b). Pod is deleted: Volume reconstruction scans pod directories for volume links
and adds the volume to asw. Can we trust the links here? Only if volume plugin
implements reconstruction correctly? What is the danger if the links are stale?
```

(*2)
`./kubernetes.io/fc/volumeDevices/28000001ff0414e2-lun-0/8576273a-ac4f-11e7-bc66-525400f59db4`

Reply:
### item2
As for the item2, we got feedback to use file descriptor at today's meeting, I'm trying to investigate the way.
From my quick test, if we keep opening(keep file descriptor as in-use) the /dev/sda file from open() systemcall during pod is alive, the /dev/sda is not reused even if the device /dev/sda is gone.

Reply:
Yeah I heard the discussion today but as per what I read symlinks in `/dev/disk/by-id` uses name of phyisical device even though symlink points to `/dev/sdxxx`. The point is, lets say I have a Samsung SSD and a WD HDD on my computer and I reboot it. Before reboot, /dev/sda is --> ssd and /dev/sdb --> is hdd. After reboot it swaps, /dev/sdb --> hdd and /dev/sda --> ssd. 

But symlinks inside `/dev/disk/by-id` will still correctly point to right device and follow the actual physical device. 

For example:

```
lrwxrwxrwx 1 root root   9 Oct  9 18:21 ata-Crucial_CT275MX300SSD1_16431465704F -> ../../sda
lrwxrwxrwx 1 root root  10 Oct  9 18:21 ata-Crucial_CT275MX300SSD1_16431465704F-part1 -> ../../sda1
lrwxrwxrwx 1 root root   9 Oct  9 18:21 ata-Samsung_SSD_850_EVO_250GB_S21NNSAG433959X -> ../../sdb
lrwxrwxrwx 1 root root  10 Oct  9 18:21 ata-Samsung_SSD_850_EVO_250GB_S21NNSAG433959X-part1 -> ../../sdb1
lrwxrwxrwx 1 root root  10 Oct  9 18:21 ata-Samsung_SSD_850_EVO_250GB_S21NNSAG433959X-part2 -> ../../sdb2
lrwxrwxrwx 1 root root   9 Oct  9 18:21 ata-Samsung_SSD_850_EVO_500GB_S2RANX0J337866Y -> ../../sdc
lrwxrwxrwx 1 root root  10 Oct  9 18:21 ata-Samsung_SSD_850_EVO_500GB_S2RANX0J337866Y-part1 -> ../../sdc1
lrwxrwxrwx 1 root root  10 Oct  9 18:21 ata-Samsung_SSD_850_EVO_500GB_S2RANX0J337866Y-part2 -> ../../sdc2
lrwxrwxrwx 1 root root  10 Oct  9 18:21 ata-Samsung_SSD_850_EVO_500GB_S2RANX0J337866Y-part5 -> ../../sdc5
lrwxrwxrwx 1 root root  10 Oct  9 18:21 ata-Samsung_SSD_850_EVO_500GB_S2RANX0J337866Y-part6 -> ../../sdc6
lrwxrwxrwx 1 root root  10 Oct  9 18:21 ata-Samsung_SSD_850_EVO_500GB_S2RANX0J337866Y-part7 -> ../../sdc7
lrwxrwxrwx 1 root root   9 Oct  9 18:21 wwn-0x5002538d41de9e27 -> ../../sdc
lrwxrwxrwx 1 root root  10 Oct  9 18:21 wwn-0x5002538d41de9e27-part1 -> ../../sdc1
lrwxrwxrwx 1 root root  10 Oct  9 18:21 wwn-0x5002538d41de9e27-part2 -> ../../sdc2
lrwxrwxrwx 1 root root  10 Oct  9 18:21 wwn-0x5002538d41de9e27-part5 -> ../../sdc5
lrwxrwxrwx 1 root root  10 Oct  9 18:21 wwn-0x5002538d41de9e27-part6 -> ../../sdc6
lrwxrwxrwx 1 root root  10 Oct  9 18:21 wwn-0x5002538d41de9e27-part7 -> ../../sdc7
lrwxrwxrwx 1 root root   9 Oct  9 18:21 wwn-0x5002538da0240e9c -> ../../sdb
lrwxrwxrwx 1 root root  10 Oct  9 18:21 wwn-0x5002538da0240e9c-part1 -> ../../sdb1
lrwxrwxrwx 1 root root  10 Oct  9 18:21 wwn-0x5002538da0240e9c-part2 -> ../../sdb2
lrwxrwxrwx 1 root root   9 Oct  9 18:21 wwn-0x500a07511465704f -> ../../sda
lrwxrwxrwx 1 root root  10 Oct  9 18:21 wwn-0x500a07511465704f-part1 -> ../../sda1
```
So I would assume that those symlinks will "follow" their physical disks.

Reply:
@gnufied 

>> So I would assume that those symlinks will "follow" their physical disks.

Yeah, this is because udev always creates/updates symlink under /dev/disk/by-id using scsi_id command like "/lib/udev/scsi_id --page=0x83 --whitelisted --device=/dev/sdX".
Issusing the scsi command page 0x83 returns unique physical disk ID and this ID is actually contained symlink name under by-id.

I may be misunderstanding. What place do you suggest to use /dev/disk/by-id?

Reply:
I assume that `/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/{pod uuid1}` points to physical device like `/dev/xvdba`, so what I was suggesting is - rather than symlinking to `/dev/xvdba`  - `/var/lib/kubelet/plugins/xxx/` could symlink to device id in `/dev/disk/by-id/xxxx` which would be guaranteed to be accurate across reboots.

Reply:
@gnufied 
Thanks. I think it works.
Right now we have following two paths(*1), device path for passing into container is obtained from the destination of Pod device path. So we need to symlink pod device path to by-id.
I suppose the steps would be like (*2).

(*1)
- Global map path
/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/{pod uuid1}

- Pod device path
/var/lib/kubelet/pods/{pod uuid}/volumeDevices/kubernetes.io~fc/block-pv0001

(*2)
1. Get devicePath from WaitForAttach() or SetUpDevice().
1. Search symlink under /dev/disk/by-id/ corresponding to devicePath if /dev/disk/by-id directory
1. If /dev/disk/by-id directory doesn't exist, search symlink under /dev/ corresponding to devicePath

Reply:
@gnufied 
I remember that there is a destination check for Global map path during GetDeviceSymlinkRefs(), so we also need to symlink Global map path to by-id if possible.

- GetDeviceSymlinkRefs(devPath string, mapPath string)